src/darwin/Framework/CHIP/templates/partials/test_cluster.zapt

```diff
@@ -25,50 +30,92 @@
     {{chipType}} {{asLowerCamelCase name}}Argument = {{definedValue}}{{asTypeLiteralSuffix chipType}};
     {{/if}}
     {{/chip_tests_item_parameters}}
-    {{/if}}
     {{#if isCommand}}
     [cluster {{asLowerCamelCase command}}:{{#chip_tests_item_parameters}}{{#not_first}}{{asLowerCamelCase name}}:{{/not_first}}{{asLowerCamelCase name}}Argument {{#last}}responseHandler:{{/last}}{{/chip_tests_item_parameters}}^(NSError * err, NSDictionary * values) {
     {{else if isSubscribeAttribute}}
-    __block bool initialReportReceived = false;
-    [cluster reportAttribute{{asUpperCamelCase attribute}}WithResponseHandler:^(NSError * err, NSDictionary * values) {
-        NSLog(@"Reporting Test Report: %@", err);
-        XCTAssertEqual(err.code, 0);
-        {{> process_response_value}}
-        initialReportReceived = true;
-    }];
-
     [cluster configureAttribute{{asUpperCamelCase attribute}}WithMinInterval:{{minInterval}}
                                                                  maxInterval:{{maxInterval}}
                                                              responseHandler:^(NSError * err, NSDictionary * values) {
+    {{else if isWaitForReport}}
+    [cluster reportAttribute{{asUpperCamelCase attribute}}WithResponseHandler:^(NSError * err, NSDictionary * values) {
     {{else if isReadAttribute}}
     [cluster readAttribute{{asUpperCamelCase attribute}}WithResponseHandler:^(NSError * err, NSDictionary * values) {
     {{else if isWriteAttribute}}
     [cluster writeAttribute{{asUpperCamelCase attribute}}WithValue:{{#chip_tests_item_parameters}}{{asLowerCamelCase name}}Argument{{/chip_tests_item_parameters}} responseHandler:^(NSError * err, NSDictionary * values) {
     {{/if}}
         NSLog(@"{{label}} Error: %@", err);
 
+{{!-- Set the flag for async tests, or fulfill the expection directly. --}}
+{{#*inline "FulfillOrSetFinishFlag"}}
+{{#unless async}}
+[expectation fulfill];
+{{else}}
+testSendCluster{{parent.filename}}_{{asTestIndex index}}_{{asUpperCamelCase command}}_Finished = true;
+{{/unless}}
+{{/inline}}
+
         {{#if optional}}
         if (err.code == CHIPErrorCodeUnsupportedAttribute) {
-            [expectation fulfill];
+            {{> FulfillOrSetFinishFlag}}
```

Comment:
I'm not sure exactly what to do for `optional` and `async`. Can we remove the change here and we will figure out what to do later ? Maybe we should throw in the generation process if both are used at the same time...